### PR TITLE
Added subscribers and other details to flightController

### PR DESCRIPTION
**Pull request by GPayne**
Fix for [OpenSpace issue#3397](https://github.com/OpenSpace/OpenSpace/issues/3397)

### CODE REVIEW

File: src/api/Middleware/flightController.js [hunk 1]
```diff
@@ -1,28 +1,82 @@
-import actionTypes from '../Actions/actionTypes';
+import { connectFlightController, disconnectFlightController, sendFlightControl } from '../Actions';
+import { actionTypes } from '../Actions/actionTypes';
+import { rootOwnerKey } from '../keys';
+
 import api from '../api';
 
-let topic;
+let topic = undefined;
+let nSubscribers = 0;
```
Comment (ylvaselling) [inline src/api/Middleware/flightController.js:8]: Removed unused imports
```
import actionTypes from '../Actions/actionTypes';
import api from '../api';

let topic = undefined;
let nSubscribers = 0;
```

File: src/api/Middleware/flightController.js [hunk 2]
```diff
@@ -1,28 +1,82 @@
...
+  if (topic) {
+    topic.talk({
+      event: 'refresh'
+    });
+  } else {
+    // If we do not have a subscription, we need to create a new topic.
+    const tempTopic = api.startTopic('flightcontroller', {
+      event: 'refresh'
+    });
+  }
```
Comment (ylvaselling) [inline src/api/Middleware/flightController.js:40]:
```suggestion
async function connectToFlightControllerTopic(store) {
  topic = api.startTopic(
    'flightcontroller',
    {
      type: "connect"
    }
  );
}

const unsubscribe = () => {
  if (!topic) {
    return;
  }
  topic.talk({
    event: 'disconnect'
  });
  topic.cancel();
}
```

Comment (ylvaselling) [reply src/api/Middleware/flightController.js:40]: There is no "stop_subscription" or "refresh" in the `flightcontrollertopic.cpp` :)

File: src/api/Middleware/flightController.js [hunk 3]
```diff
@@ -1,28 +1,82 @@
...
+      break;
+    case actionTypes.connectFlightController:
+      ++nSubscribers;
+      if (nSubscribers === 1 && state.connection.isConnected) {
+        connectToFlightControllerTopic();
+      }
+      break;
+    case actionTypes.refreshFlightController:
+      refresh();
       break;
```
Comment (ylvaselling) [inline src/api/Middleware/flightController.js:65]:
```suggestion
```

File: src/api/Middleware/flightController.js [hunk 4]
```diff
@@ -1,28 +1,82 @@
...
+
 import api from '../api';
 
-let topic;
+let topic = undefined;
+let nSubscribers = 0;
 
-async function connectToFlightControllerTopic() {
-  topic = api.startTopic('flightcontroller', { type: 'connect' });
+async function connectToFlightControllerTopic(store) {
```
Comment (WeirdRubberDuck) [inline src/api/Middleware/flightController.js:10]:
```suggestion
async function connectToFlightControllerTopic() {
```

File: src/api/Middleware/flightController.js [hunk 5]
```diff
@@ -1,28 +1,82 @@
...
+    });
+  }
 }
 
 function sendFlightControlMessage(data) {
   topic.talk(data);
 }
 
-const flightController = () => (next) => (action) => {
+export const flightController = store => next => action => {
```
Comment (WeirdRubberDuck) [inline src/api/Middleware/flightController.js:47]:
```suggestion
export const flightController = (store) => (next) => (action) => {
```
To comply with out linting settings

File: src/api/Middleware/flightController.js [hunk 6]
```diff
@@ -1,28 +1,82 @@
-import actionTypes from '../Actions/actionTypes';
+import { connectFlightController, disconnectFlightController, sendFlightControl } from '../Actions';
+import { actionTypes } from '../Actions/actionTypes';
+import { rootOwnerKey } from '../keys';
+
 import api from '../api';
 
-let topic;
+let topic = undefined;
```
Comment (WeirdRubberDuck) [inline src/api/Middleware/flightController.js:7]:
```suggestion
let topic;
```

This does not have to be initialized as undefined, and our linting currently warns about it.

File: src/api/Middleware/flightController.js [hunk 7]
```diff
@@ -1,28 +1,82 @@
-import actionTypes from '../Actions/actionTypes';
+import { connectFlightController, disconnectFlightController, sendFlightControl } from '../Actions';
+import { actionTypes } from '../Actions/actionTypes';
```
Comment (WeirdRubberDuck) [inline src/api/Middleware/flightController.js:2]:
```suggestion
import actionTypes from '../Actions/actionTypes';
```
This currently does not work

**Review by WeirdRubberDuck — CHANGES_REQUESTED** [review]
I'm quite unaware of the context for this PR, so can't say much about why the subscribers are needed or comment on that part. But it seems to me like it does not actually fix the issue https://github.com/OpenSpace/OpenSpace/issues/3397? I managed to reproduce it even with this update, by following the steps in the issue

Otherwise, I added some comments to make the code comply to our eslint settings. I recommend installing the ESLint plugin for VS Code, to more easily catch these errors during the coding

**Comment by GPayne** [comment]
Retracting this PR because my test results were a false positive. I'm unable to reproduce the fix, probably due to my original test setup being flawed somehow.
A rewrite of this frontend code is currently in progress, and [this issue](https://github.com/OpenSpace/OpenSpace/issues/3397) may well be fixed as part of that effort.

**Pull request closed without merging**